### ISSUE TITLE
Server-side rendering uses blocking <script> tags having a negative impact to page load performance

### ISSUE THREAD

**Issue by miohtama** [sveltejs/kit#3102]
### Describe the problem

Currently, SvelteKit server-side rendering puts all loaded JavaScript to `<head>` section of the page. This negatively affects the user-perceived performance, as First Contentful Paint (FCP) is delayed because the browser spends time loading often heavy .js files.

This is how the JavaScript loading looks in the server-side rendered page. It is using [modulepreload](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/modulepreload) that is blocking:

```html
<head>
<link rel="modulepreload" href="/_app/chunks/vendor-87f5cdbf.js">
</head>
```

Because the full HTML page is rendered on the server-side in any case, there is no reason for blocking the load with `<script>` within `<head>`.

[A more detailed breakdown of <script> tags in SvelteKit can be found in this StackOverflow question](https://stackoverflow.com/questions/70471512/analyzing-optimizing-and-lazy-loading-vendor-js-when-doing-sveltekit-server-sid).

[Here is the related Vite issue](https://github.com/vitejs/vite/issues/5120).

### Describe the proposed solution

The JavaScript only needs to load after the page is hydrated. For non-hydrated pages, no JavaScript is needed at all.

* When the server-side page is rendered use `<script defer async>` instead of `<link rel="modulepreload">` for the JavaScript. [defer attribute is documented here](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script)

### Alternatives considered

Alternatively, all JavaScript can be loaded before closing `</body>`

* Break down `app.html` template to separate `svelte.head` and `svelte.scripts` sections.
* Move all JavaScript loading to `svelte.scripts` that is placed before `</body>` 

### Importance

would make my life easier

### Additional Information

SvelteKit should ship with very optimised server-side rendering settings, as this will positively affect the perceived quality of SvelteKit and teach newcomers the best way of doing server-side rendered pages.

Some client-side JavaScript may be required to be loaded in `head`. However, this is more of an exception on nowadays and such JavaScript is unlikely to work with server-side rendering in any case. 

For such JavaScript files and bundles, the existing client-side only JavaScript loading should be applied.

[Here is someone's workaround that he post-processes HTML and strips away modulepreload](https://github.com/vitejs/vite/issues/5120#issuecomment-950122002).

**Comment by miohtama** [comment]
[Here is my writeup how to transform generated HTML in SvelteKit hooks using cheetah](https://stackoverflow.com/questions/70471512/analyzing-optimizing-and-lazy-loading-vendor-js-when-doing-sveltekit-server-sid/70479348#70479348). By changing `<link type=modulepreload>` to the end of `<body>` `<script>` tags I managed to reduce the page loading time significantly (4.0 -> 2.5 to FCP on mobile).

**Comment by benmccann** [comment]
> It is using modulepreload that is blocking

Can you provide a source for that? It's not my understanding that it is blocking.

> Because the full HTML page is rendered on the server-side in any case, there is no reason for blocking the load with `<script>` within `<head>`.

You used a `<link>` tag in the sentence before this one. Did you mean to write `<link>` here? If you correctly wrote `<script>` then I'm confused  by the example you gave which doesn't have any `<script>` tag

> The JavaScript only needs to load after the page is hydrated.

That's incorrect. The JavaScript is what does the hydration, so it would have to load before the page is hydrated

> When the server-side page is rendered use <script defer async> instead of <link rel="modulepreload"> for the JavaScript. defer attribute is documented here

`modulepreload` only fetches the script from the server so that they're available when needed. It doesn't block the rendering of the current page. Furthermore, modules are automatically deferred: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules#other_differences_between_modules_and_standard_scripts

**Comment by miohtama** [comment]
> modulepreload only fetches the script from the server so that they're available when needed. It doesn't block the rendering of the current page. Furthermore, modules are automatically deferred: 

[If you check the underlying Vite issue](https://github.com/vitejs/vite/issues/5120) it looks like this assumption is incorrect.

**Comment by benmccann** [comment]
I responded on that issue: https://github.com/vitejs/vite/issues/5120#issuecomment-1001688323. I'm unconvinced that this is a good idea.

I found your post on StackOverflow. I think the bigger issue in your case is the unnecessary size of the vendor chunk. Can you try removing the `modulepreload` workaround and disabling the vendor chunk instead? https://stackoverflow.com/a/70499134/64421

Let me know if that helps. I'd be interested in trying to change the Vite default there.

**Comment by miohtama** [comment]
It does not help, because JS is blocking the page load, regardless of what's the JavaScript size.

**Comment by benmccann** [comment]
> JS is blocking the page load

I don't think it really is in any meaningful way. The issue you linked to says "CSS files are blocking page rendering". It then goes on to state that while the JS and CSS are loaded in parallel, Chrome is kicking off the requests for the JS files first adding an extra 0.5ms delay for each JS file that's preloaded. You've got about 10 JS files on https://tradingstrategy.ai/, so that'd be about a 5ms delay. That's pretty negligible.

This is hard to understand without an example I can see. Is there some way you can share a project with me that reproduces this. Or could you at deploy a version without your workarounds so that I can see what you were seeing that you were displeased with?

**Comment by miohtama** [comment]
Thank you @benmccann  Let me get back to you with a repeatable sample. Might take some days.

**Comment by miohtama** [comment]
@benmccann Based on your comments, I did some quick tests on my local dev with Chrome performance tab + 3G throttling. Here is the original "bad" version with 10s FCP on mobile, without any optimizations done. You can see that CSS load is not immediately triggered.

<img width="968" alt="image" src="https://user-images.githubusercontent.com/49922/147500988-c4f24312-49d8-4fa7-b386-acb85b601140.png">

Here is a a version with `<link>` -> `<script>` transformation, FCP 5 seconds. You can see that the FCP is around the time `_layout.svelte.css` bundle finishes loading. You can see CSS files start load instantly after HTML is available.

<img width="1060" alt="image" src="https://user-images.githubusercontent.com/49922/147501180-18a41afa-e611-41ab-9e43-89551bfa394a.png">

Here is the optimized version with `<link type="modulepreload>` left in intact- FCP around 10s. You can see that all CSS is not loaded immediately. 

<img width="1468" alt="image" src="https://user-images.githubusercontent.com/49922/147501491-e9087bf7-16a6-4e65-ada4-35931a5bf813.png">

From the screenshot above you can see that Chrome, for some reason, delays the start of loading of CSS files and processes some Svelte JS files first. `fonts.css` loading at the top of the head directly from `/static` seems suspicious....

Just for the sake of it, I made one version where I moved `fonts.css` to the bottom of the `<head>` to see if it causes the issue, but still getting FCP around 10 seconds mark.

<img width="1075" alt="image" src="https://user-images.githubusercontent.com/49922/147501732-0b783fb7-a448-4c37-b155-f26a4cea4d3d.png">

And to make sure I am not insane, or missing something else, and there is some other issue affecting the loading times, I repeated the test run one more time with `<link>` -> `<script>` transformation and this time having `fonts.css` at the bottom of `<head>` getting FCP 7 seconds:

<img width="1199" alt="image" src="https://user-images.githubusercontent.com/49922/147501915-1f6b05f7-1baa-4bd3-918e-99a7fa30427c.png">

One reason I think might affect the behaviour is the bandwidth available on mobile connections. If the web browser is preloading JS files, it will eat away bandwidth that would be otherwise available to load CSS files.

**Comment by benmccann** [comment]
This is very different than the behavior I've ever seen on any other SvelteKit site. I always see CSS fetched immediately. E.g. here is the Network tab with slow 3g throttling on svelte.dev

![Screenshot from 2021-12-27 13-28-28](https://user-images.githubusercontent.com/322311/147508535-0c6f9622-1179-4e74-850b-2be6d3e36f86.png)

I wonder why things would be different in your case. If you could share a repository with me to reproduce this that would be ideal

**Comment by benmccann** [comment]
Fixed in https://github.com/sveltejs/kit/commit/d138efe21692f5925f1e89afc0a33f42d6a1a711

**Comment by miohtama** [comment]
You are the man @benmccann 🙌  - I will try the patched Github master now.

**Comment by miohtama** [comment]
On a side note, what's the easiest way to test SvelteKit master in a production? Can I do an npm install against a specific Github commit easily?

**Comment by benmccann** [comment]
It's already been released. Just upgrade your version of SvelteKit

**Comment by miohtama** [comment]
I can confirm this is fixed. The CSS loads now first, then JS and the First Contentful Paint (FCP) happens as soon as possible.

<img width="875" alt="image" src="https://user-images.githubusercontent.com/49922/148426905-99ffb1d3-f4a7-4525-8d76-407a0686029e.png">
